### ISSUE TITLE
False positive in `keyword_idents_2024`

### ISSUE THREAD

**Issue by jhpratt** [rust-lang/rust#124862]
```rust
#![warn(keyword_idents_2024)]

#[allow(unused)]
macro_rules! foo {
    ($gen:expr) => {
        $gen
    };
}
```

results in a warning

```text
warning: `gen` is a keyword in the 2024 edition
 --> src/main.rs:5:7
  |
5 |     ($gen:expr) => {
  |       ^^^ help: you can use a raw identifier to stay compatible: `r#gen`
  |
  = warning: this is accepted in the current edition (Rust 2021) but is a hard error in Rust 2024!
  = note: for more information, see issue #49716 <https://github.com/rust-lang/rust/issues/49716>
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(keyword_idents_2024)]
  |         ^^^^^^^^^^^^^^^^^^^

warning: `gen` is a keyword in the 2024 edition
 --> src/main.rs:6:10
  |
6 |         $gen
  |          ^^^ help: you can use a raw identifier to stay compatible: `r#gen`
  |
  = warning: this is accepted in the current edition (Rust 2021) but is a hard error in Rust 2024!
  = note: for more information, see issue #49716 <https://github.com/rust-lang/rust/issues/49716>
```

The suggested code _does_ compile (to my surprise), but the warning is a false positive. Switching the edition to 2024 results in a successful compilation as written. This is consistent with all other keywords that can be used as matchers, so I can't imagine this isn't intended behavior.

rustc info:

```text
rustc 1.80.0-nightly (7d83a4c13 2024-05-06)
binary: rustc
commit-hash: 7d83a4c131ab9ae81a74c6fd825c827d74a2881d
commit-date: 2024-05-06
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4
```

**Comment by compiler-errors** [comment]
This was preexisting in the lint (fires for `$async` too in ed 2015), but perhaps more likely to be encountered with the `$gen` fragment matcher. I can fix it.

**Comment by jhpratt** [comment]
I was just looking into it myself. I think it's a two line fix? Copying what's done for `dyn` seems like it should work.

https://github.com/rust-lang/rust/pull/123680/files#diff-f16097779719dd3638857eefc5ee3eb5052f3d1921df2c672153601aa5f5242fR1906

**Comment by compiler-errors** [comment]
@jhpratt:

That's too heavy of a hammer, and will cause this test to *not* fail even though it should:

```
#![deny(keyword_idents_2024)]

macro_rules! t {
    () => { fn gen() {} }
}

t!();
```

**Comment by compiler-errors** [comment]
We should specifically be detecting macro vars that have keyword names instead.

**Comment by compiler-errors** [comment]
But if you're looking into it already, be my guest. Just make sure to claim it next time :)

**Comment by jhpratt** [comment]
Nah, go ahead! I merely saw the PR that added the lint and noticed something that might have been of interest.

**Comment by compiler-errors** [comment]
Alright then :+1: I'll reclaim
